xls/modules/zstd/buffer.x

```diff
@@ -0,0 +1,424 @@
+    content: bits[CAPACITY],
+    length: u32
+}
+
+// Status values reported by the functions operating on a Buffer
+pub enum BufferStatus : u2 {
+    OK = 0,
+    NO_ENOUGH_SPACE = 1,
+    NO_ENOUGH_DATA = 2,
+}
+
+// Structure for returning Buffer and BufferStatus together
+pub struct BufferResult<CAPACITY: u32> {
+    buffer: Buffer<CAPACITY>,
+    status: BufferStatus
+}
+
+// Checks whether a `buffer` can fit `data`
+pub fn buffer_can_fit<CAPACITY: u32, DSIZE: u32>(buffer: Buffer<CAPACITY>, data: bits[DSIZE]) -> bool {
+    buffer.length + DSIZE <= CAPACITY
+}
+
+#[test]
+fn test_buffer_can_fit() {
+    let buffer = Buffer<u32:32> { content: u32:0, length: u32:0 };
+    assert_eq(buffer_can_fit(buffer, bits[0]:0), true);
+    assert_eq(buffer_can_fit(buffer, u16:0), true);
+    assert_eq(buffer_can_fit(buffer, u32:0), true);
+    assert_eq(buffer_can_fit(buffer, u33:0), false);
+
+    let buffer = Buffer<u32:32> { content: u32:0, length: u32:16 };
+    assert_eq(buffer_can_fit(buffer, bits[0]:0), true);
+    assert_eq(buffer_can_fit(buffer, u16:0), true);
+    assert_eq(buffer_can_fit(buffer, u17:0), false);
+    assert_eq(buffer_can_fit(buffer, u32:0), false);
+
+    let buffer = Buffer<u32:32> { content: u32:0, length: u32:32 };
+    assert_eq(buffer_can_fit(buffer, bits[0]:0), true);
+    assert_eq(buffer_can_fit(buffer, u1:0), false);
+    assert_eq(buffer_can_fit(buffer, u16:0), false);
+    assert_eq(buffer_can_fit(buffer, u32:0), false);
+}
+
+// Checks whether a `buffer` has at least `length` amount of data
+pub fn buffer_has_at_least<CAPACITY: u32>(buffer: Buffer<CAPACITY>, length: u32) -> bool {
+    length <= buffer.length
+}
+
+#[test]
+fn test_buffer_has_at_least() {
+    let buffer = Buffer { content: u32:0, length: u32:0 };
+    assert_eq(buffer_has_at_least(buffer, u32:0), true);
+    assert_eq(buffer_has_at_least(buffer, u32:16), false);
+    assert_eq(buffer_has_at_least(buffer, u32:32), false);
+    assert_eq(buffer_has_at_least(buffer, u32:33), false);
+
+    let buffer = Buffer { content: u32:0, length: u32:16 };
+    assert_eq(buffer_has_at_least(buffer, u32:0), true);
+    assert_eq(buffer_has_at_least(buffer, u32:16), true);
+    assert_eq(buffer_has_at_least(buffer, u32:32), false);
+    assert_eq(buffer_has_at_least(buffer, u32:33), false);
+
+    let buffer = Buffer { content: u32:0, length: u32:32 };
+    assert_eq(buffer_has_at_least(buffer, u32:0), true);
+    assert_eq(buffer_has_at_least(buffer, u32:16), true);
+    assert_eq(buffer_has_at_least(buffer, u32:32), true);
+    assert_eq(buffer_has_at_least(buffer, u32:33), false);
+}
+
+// Returns a new buffer with `data` appended to the original `buffer`.
+// It will fail if the buffer cannot fit the data. For calls that need better
+// error handling, check `buffer_append_checked`
+pub fn buffer_append<CAPACITY: u32, DSIZE: u32> (buffer: Buffer<CAPACITY>, data: bits[DSIZE]) -> Buffer<CAPACITY> {
+    if buffer_can_fit(buffer, data) == false {
```

Comment:
q: should we get rid of the check here, since it is supposed to be the unchecked version?

Reply:
Checked/unchecked variants refer to the way the failure is reported in resulting verilog.
For the unchecked variant this check only indicates that something went wrong at the IR simulation level. Codegen verilog for that won't have any logic for that.
I think it is good idea to keep this check to signal the user potential issues in his/hers DSLX logic

Reply:
since the `fail!` built-in emit as `$fatal` in verilog I do think the associated logic for compute the condition will also get codegen'ed (but maybe optimized away by synthesis?)

rather than having two variant of the function, maybe we could only keep the unchecked one (w/ the `fail!`) and have the caller rely on calling `can_fit` if they need to do something dynamic?

wdyt?

Reply:
As far as I remember `$fatal` is available only in SystemVerilog. For our benchmarks we do codegen with Verilog which does not generate any logic for this. On the other hand, it generates `$display()` for `trace_fmt!()` and those are then removed in the synthesis process so it is possible that logic for `$fatal` would also be removed.

> rather than having two variant of the function, maybe we could only keep the unchecked one (w/ the fail!) and have the caller rely on calling can_fit if they need to do something dynamic?

This is good idea, we can probably do that.

Reply:
On the other hand when I looked closely at the implementation after removing `_checked` variants of the functions it does not look that good. For example, we use `_checked` functions in frame header decoding in 4 places in single `frame_header.x` file. Removing `_checked` functions would require placing checks straight in the logic of frame header decoder. It would look something like that for each usage of `_checked` variant of the functions:
```
diff --git a/xls/modules/zstd/frame_header.x b/xls/modules/zstd/frame_header.x
index 785586396..01cd927cd 100644
--- a/xls/modules/zstd/frame_header.x
+++ b/xls/modules/zstd/frame_header.x
@@ -115,7 +115,24 @@ fn test_extract_frame_header_descriptor() {
 // returns BufferResult with the outcome of the operations on the buffer and
 // information extracted from the Frame_Header_Descriptor
 fn parse_frame_header_descriptor<CAPACITY: u32>(buffer: Buffer<CAPACITY>) -> (BufferResult<CAPACITY>, FrameHeaderDescriptor) {
-    let (result, data) = buff::buffer_fixed_pop_checked<CAPACITY, u32:8>(buffer);
+    let (result, data) = if (buff:buffer_has_at_least(buffer, u32:8)) {
+        let (buffer_leftover, content) = buff::buffer_fixed_pop<CAPACITY, u32:8>(buffer);
+        (
+            buff::BufferResult {
+                status: buff:BufferStatus::OK,
+                buffer: buffer_leftover
+            },
+            content
+        )
+    } else {
+        (
+            buff::BufferResult {
+                status: buff::BufferStatus:NO_ENOUGH_DATA,
+                buffer: buffer
+            },
+            bits[CAPACITY]:0
+        )
+    };
     match result.status {
         BufferStatus::OK => {
             let frame_header_desc = extract_frame_header_descriptor(data);
```

It is possible to put the code from `match` straight under the `if` but this will still lead to unnecessary code duplication. Maybe it would be better to keep those `_checked` functions?

Reply:
I'd be curious to see what the calling logic does when handling `buff::BufferStatus:NO_ENOUGH_DATA`, can you point me to it?

My reasoning is that if you're gonna act on something related to the size of the buffer, you'd rather do this early on to make it explicit how you handle this case from the protocol pov, rather than relying on propagating the status back from the buffer call.

Reply:
OK, so the part that is responsible for handling `NO_ENOUGH_DATA` is the top level proc of the `ZstdDecoder`. The code for this is not a part of any open PR yet as we still work on fixing bugs found in our C++ tests where we compare our decoder against `libzstd`.
Top level proc has states which mark the stage of frame decoding, e.g.:
```
enum ZstdDecoderStatus : u8 {
  DECODE_MAGIC_NUMBER = 0,
  DECODE_FRAME_HEADER = 1,
  DECODE_BLOCK_HEADER = 2,
  FEED_BLOCK_DECODER = 3,
  DECODE_CHECKSUM = 4,
  ERROR = 255,
}
```
When `NO_ENOUGH_DATA` occurs we just stay in the current state for the next evaluation where we receive data from decoder input and append it to the buffer. Then we can try again to decode the part specified by the state.

Reply:
We've opened a PR which contains the mentioned code. It's https://github.com/google/xls/pull/1315. You can check the example of handling `NO_ENOUGH_DATA` [here](https://github.com/google/xls/pull/1315/files#diff-cb74ede7463d11191318780b775a62a1139076d590ea087065aaefe167a36c4fR114).